modules/boulder/src/main/ts/ephox/boulder/core/StructureProcessor.ts

```diff
@@ -0,0 +1,275 @@
+import { Arr, Fun, Merger, Obj, Optional, Thunk, Type } from '@ephox/katamari';
+import { SimpleResult, SimpleResultType } from '../alien/SimpleResult';
+import * as FieldPresence from '../api/FieldPresence';
+import * as Objects from '../api/Objects';
+import { ResultCombine } from '../combine/ResultCombine';
+import * as FieldProcessor from './FieldProcessor';
+import * as ObjWriter from './ObjWriter';
+import * as SchemaError from './SchemaError';
+import { value } from './Utils';
+
+type FieldProcessor = FieldProcessor.FieldProcessor;
+
+type SchemaError = SchemaError.SchemaError;
+
+export type ValueValidator = (a) => SimpleResult<string, any>;
+export type PropExtractor = (path: string[], val: any) => SimpleResult<SchemaError[], any>;
+export interface StructureProcessor {
+  readonly getProp: PropExtractor;
+  readonly toString: () => string;
+}
+
+const output = (newKey: string, value: any): FieldProcessor => FieldProcessor.customField(newKey, Fun.constant(value));
+
+const snapshot = (newKey: string): FieldProcessor => FieldProcessor.customField(newKey, Fun.identity);
+
+const strictAccess = <T>(path: string[], obj: Record<string, T>, key: string): SimpleResult<SchemaError[], T> => {
+  // In strict mode, if it undefined, it is an error.
+  return Obj.get(obj, key).fold<SimpleResult<SchemaError[], any>>(() =>
+    SchemaError.missingStrict(path, key, obj), SimpleResult.svalue);
+};
+
+const fallbackAccess = <T>(obj: Record<string, T>, key: string, fallbackThunk: (obj: Record<string, T>) => T): SimpleResult<SchemaError[], T> => {
+  const v = Obj.get(obj, key).fold(() => fallbackThunk(obj), Fun.identity);
+  return SimpleResult.svalue(v);
+};
+
+const optionAccess = <T>(obj: Record<string, T>, key: string): SimpleResult<SchemaError[], Optional<T>> =>
+  SimpleResult.svalue(Obj.get(obj, key));
+
+const optionDefaultedAccess = <T>(obj: Record<string, T | true>, key: string, fallback: (obj: Record<string, T | true>) => T): SimpleResult<SchemaError[], Optional<T>> => {
+  const opt = Obj.get(obj, key).map((val) => val === true ? fallback(obj) : val);
+  return SimpleResult.svalue(opt);
+};
+
+type SimpleBundle = SimpleResult<SchemaError[], any>;
```

Comment:
Can we move these to the top please? We don't normally have types/interfaces in the middle of a file.